### DISCUSSION TITLE
Ability to implicitly set a default filename through environment variables

### DISCUSSION THREAD

**Discussion by hovsater** [jotaen/klog#203] [Feature ideas & discussions]
I tend to have a one `.klg` file per project to track the amount of time I've spent on it. While bookmarks is a nice feature, it's also kind of repetitive having to specify it over and over again. It would be nice if you could implicitly tell klog what file to use through an environment variable. This goes really well together with [direnv](https://direnv.net).

This would allow me to create a `.envrc` file in the root of my project, add `export KLOG_FILE=./proj.klg` to it and let [direnv](https://direnv.net) automatically export it for me. klog would then automatically pick up the `KLOG_FILE` environment variableif it were present.

What do you think?

**Comment by jotaen** [comment]
Thanks for the suggestion, I didn’t know about direnv.

I have to admit, I’m slightly reluctant to add more ways to specify input, in order to keep the complexity contained. I wouldn’t rule it out, but I probably have to think more about it.

> it's also kind of repetitive having to specify it over and over again

Do you mean it’s repetitive to type the bookmark name when using it, or to set it up in the first place? Personally, I always assign two-letter bookmark names, e.g. `@hw` for `housework.klg`. ~I’m also hoping for klog to support [shell completion](https://github.com/jotaen/klog/issues/177) soon, which would help to speed up the typing as well. (Unfortunately, that’s [currently blocked on an issue with a third-party library](https://github.com/WillAbides/kongplete/issues/15).)~ **UPDATE: [Tab completion has landed in klog v5.2](https://github.com/jotaen/klog/blob/main/CHANGELOG.md#v52)**

**Reply by hovsater** [reply]
> Do you mean it’s repetitive to type the bookmark name when using it, or to set it up in the first place? Personally, I always assign two-letter bookmark names, e.g. @hw for housework.klg

Yes. I'm working on multiple client projects and I use klog to track my time across different files (on per client). I have a bookmark for each client, but it's rather tedious having to repeat the bookmark over and over.

Perhaps we could improve the default bookmark functionality instead? Currently, you don't have to specify a bookmark if there's a special `@default` bookmark present. What if we could adjust the default bookmark to use with an environment variable such as `KLOG_DEFAULT_BOOKMARK`? This would solve my use case as I could use [direnv](https://direnv.net) to explicitly set that environment variable to my client bookmark on a per-project basis.

I could even submit a pull request for it, if you think it's worth adding. 🙂

**Reply by hovsater** [reply]
I think it would be as easy as amending the following:

https://github.com/jotaen/klog/blob/592ebb9d64295981c8e6ac47a2345f4a6da850c0/klog/app/bookmark.go#L19-L24

to make use of `os.LookupEnv` and fallback to `BOOKMARK_DEFAULT_NAME` in case it's not set.

**Reply by jotaen** [reply]
As outlined in https://github.com/jotaen/klog/pull/269, I think `bookmark.go` isn’t the right place for this. Right now, and with klog’s current code design, I don’t see any quick-and-simple approach to make this feature request happen, unfortunately…

Another workaround for you that would come to my mind is this: you could try to leverage the fact that the klog config folder can be specified via the [`KLOG_CONFIG_HOME` environment variable](https://github.com/jotaen/klog/blob/592ebb9d64295981c8e6ac47a2345f4a6da850c0/klog.go#L94C30-L94C30). So in your direnv file, you could set `KLOG_CONFIG_HOME` to be e.g. `/home/my-project-123/klog`. That way, klog would look inside this folder for the `bookmarks.json` file, and hence use the default bookmark from there. In case you maintain multiple `.klg` files per project (e.g., one per month), you could also put them all inside this local `klog/` folder, next to `bookmarks.json`. You can also name that folder differently, e.g., `time-tracking/` instead of `klog/`.

This approach has a one downside, however: the project-specific klog-folder is independent of the “global” one (e.g., `~/.klog`). So inside your respective direnv scope, you’d (1) lose access to all other “global” bookmarks, unless you manually mirror them, and (2) in case you had created a global `config.ini` configuration file, you’d have to either manually mirror or symlink it into the project-specific klog folder.

**Reply by hovsater** [reply]
Thanks for the response. I'll just let this be for now. 👍
